validator/rpc/handlers_keymanager.go

```diff
@@ -19,11 +21,289 @@ import (
+	if err != nil {
+		http2.HandleError(w, errors.Wrap(err, "Could not import keystores").Error(), http.StatusInternalServerError)
+		return
+	}
+
+	// If any of the keys imported had a slashing protection history before, we
+	// stop marking them as deleted from our validator database.
+	http2.WriteJson(w, &ImportKeystoresResponse{Data: statuses})
+}
+
+// DeleteKeystores allows for deleting specified public keys from Prysm.
+func (s *Server) DeleteKeystores(w http.ResponseWriter, r *http.Request) {
+	ctx, span := trace.StartSpan(r.Context(), "validator.keymanagerAPI.DeleteKeystores")
+	defer span.End()
+
+	if s.validatorService == nil {
+		http2.HandleError(w, "Validator service not ready.", http.StatusServiceUnavailable)
+		return
+	}
+	if !s.walletInitialized {
+		http2.HandleError(w, "Prysm Wallet not initialized. Please create a new wallet.", http.StatusServiceUnavailable)
+		return
+	}
+	km, err := s.validatorService.Keymanager()
+	if err != nil {
+		http2.HandleError(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	var req DeleteKeystoresRequest
+	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
+		http2.HandleError(w, "Could not decode request body: "+err.Error(), http.StatusBadRequest)
+		return
+	}
+	if len(req.Pubkeys) == 0 {
+		http2.WriteJson(w, &DeleteKeystoresResponse{Data: make([]*keymanager.KeyStatus, 0)})
+		return
+	}
+	deleter, ok := km.(keymanager.Deleter)
+	if !ok {
+		sts := make([]*keymanager.KeyStatus, len(req.Pubkeys))
+		for i := 0; i < len(req.Pubkeys); i++ {
+			sts[i] = &keymanager.KeyStatus{
+				Status:  keymanager.StatusError,
+				Message: fmt.Sprintf("Keymanager kind %T cannot delete local keys", km),
+			}
+		}
+		http2.WriteJson(w, &DeleteKeystoresResponse{Data: sts})
+		return
+	}
+	bytePubKeys := make([][]byte, len(req.Pubkeys))
+	for i, pubkey := range req.Pubkeys {
+		key, ok := shared.ValidateHex(w, "Pubkey", pubkey, fieldparams.BLSPubkeyLength)
+		if !ok {
+			return
+		}
+		bytePubKeys[i] = key
+	}
+	statuses, err := deleter.DeleteKeystores(ctx, bytePubKeys)
+	if err != nil {
+		http2.HandleError(w, errors.Wrap(err, "Could not delete keys").Error(), http.StatusInternalServerError)
+		return
+	}
+
+	statuses, err = s.transformDeletedKeysStatuses(ctx, bytePubKeys, statuses)
+	if err != nil {
+		http2.HandleError(w, errors.Wrap(err, "Could not transform deleted keys statuses").Error(), http.StatusInternalServerError)
+		return
+	}
+
+	exportedHistory, err := s.slashingProtectionHistoryForDeletedKeys(ctx, bytePubKeys, statuses)
+	if err != nil {
+		log.WithError(err).Warn("Could not get slashing protection history for deleted keys")
+		sts := make([]*keymanager.KeyStatus, len(req.Pubkeys))
+		for i := 0; i < len(req.Pubkeys); i++ {
+			sts[i] = &keymanager.KeyStatus{
+				Status:  keymanager.StatusError,
+				Message: "Non duplicate keys that were existing were deleted, but could not export slashing protection history",
```

Comment:
This error message is incredibly confusing. This was my best guess of the meaning
```suggestion
				Message: "Could not export slashing protection history as existing non duplicate keys were deleted",
```